lib/Auth/Source/Twitter.php

```diff
@@ -82,92 +71,98 @@ public function __construct(array $info, array $config)
-        $parameters = [];
+        $requestToken = unserialize($state['authtwitter:authdata:requestToken']);
 
-        if (!isset($_REQUEST['oauth_token'])) {
+        $oauth_token = $request->get('oauth_token');
+        if ($oauth_token === null) {
             throw new Error\BadRequest("Missing oauth_token parameter.");
         }
-        if ($requestToken->key !== (string) $_REQUEST['oauth_token']) {
+
+        if ($requestToken->getIdentifier() !== $oauth_token) {
             throw new Error\BadRequest("Invalid oauth_token parameter.");
         }
 
-        if (!isset($_REQUEST['oauth_verifier'])) {
+        $oauth_verifier = $request->get('oauth_verifier');
+        if ($oauth_verifier === null) {
             throw new Error\BadRequest("Missing oauth_verifier parameter.");
         }
-        $parameters['oauth_verifier'] = (string) $_REQUEST['oauth_verifier'];
 
-        $consumer = new Consumer($this->key, $this->secret);
-
-        Logger::debug("oauth: Using this request token [" .
-            $requestToken->key . "] with the secret [" . $requestToken->secret . "]");
+        $server = new TwitterServer(
+            [
+                'identifier' => $this->key,
+                'secret' => $this->secret,
+            ]
+        );
 
-        // Replace the request token with an access token
-        $accessToken = $consumer->getAccessToken(
-            'https://api.twitter.com/oauth/access_token',
+        $tokenCredentials = $server->getTokenCredentials(
             $requestToken,
-            $parameters
+            $request->get('oauth_token'),
+            $request->get('oauth_verifier')
         );
-        Logger::debug("Got an access token from the OAuth service provider [" .
-            $accessToken->key . "] with the secret [" . $accessToken->secret . "]");
 
-        $verify_credentials_url = 'https://api.twitter.com/1.1/account/verify_credentials.json';
-        if ($this->include_email) {
-            $verify_credentials_url = $verify_credentials_url . '?include_email=true';
-        }
-        $userdata = $consumer->getUserInfo($verify_credentials_url, $accessToken);
+        $state['token_credentials'] = serialize($tokenCredentials);
+        $userdata = $server->getUserDetails($tokenCredentials);
 
-        if (!isset($userdata['id_str']) || !isset($userdata['screen_name'])) {
-            throw new Error\AuthSource(
-                $this->authId,
-                'Authentication error: id_str and screen_name not set.'
-            );
+        $attributes = [];
+
+        foreach ($userdata->getIterator() as $key => $value) {
+            if (is_string($value)) {
+                $attributes['twitter.' . $key] = [$value];
+            }
         }
 
-        $attributes = [];
-        foreach ($userdata as $key => $value) {
+        foreach ($userdata->extra as $key => $value) {
             if (is_string($value)) {
                 $attributes['twitter.' . $key] = [$value];
```

Comment:
No chance on collisions with the previous loop? Again would benefit from perhaps even some short explanation about the two loops

Reply:
The [API](https://github.com/thephpleague/oauth1-client/blob/master/src/Server/User.php) is a bit crappy here.. I've fixed it as good as I possibly can.. We get so much details from Twitter (like tweets) that are sent as an array, but i don't think they're relevant as an attribute.. We deal with strings and integers, the rest is discarded (for now)

Reply:
Makes sense to just ignore them (not store in state). We are interested in information likely useful for an "auth source" indeed. We'll hear from people when they should ever need them.

Reply:
That leaves us with the following attributes;

![afbeelding](https://user-images.githubusercontent.com/841045/132050052-8aa3f43f-1268-4eb9-9f8c-1166fab4fa57.png)

Reply:
Seems all you’d need?